src/Dhall/Import.hs

```diff
@@ -718,6 +737,15 @@ loadStaticWith from_import ctx n (Embed import_) = do
                 else throwM (Imported (import_:imports) (HashMismatch {..}))
 
     return expr
+loadStaticWith from_import ctx n (ImportAlt a b) = do
+  resolveExpr a `catch` handler
+    where
+      resolveExpr expr = fmap join (traverse process expr)
```

Comment:
You can simplify this to:

```haskell
resolveExpr = loadStaticWith from_import ctx n
```

Reply:
Oh right, thanks!

However, that new branch that pattern-matches on `ImportAlt` never gets called, because [the last one](https://github.com/dhall-lang/dhall-haskell/pull/473/files#diff-af3f1fd4c90ee30ad5bc7e05d5bbb9d0R749) is called on the complete expression, and that `traverse`s the tree calling always the first branch on every `Import`.

This happens because the `process` function is something like `Import -> IO (Expr Src X)`, and because the (auto-derived) implementation of `traverse` for the `ImportAlt` constructor is:
```haskell
traverse f (ImportAlt a b) = ImportAlt <$> traverse f a <*> traverse f b
```
while to get rid of this constructor and keeping the first branch that doesn't fail we would need:
```haskell
traverse f (ImportAlt a b) = 
  catch (traverse f a)
        (\e -> let _ = (e :: SomeException) in traverse f b)
```

So basically I think that the simplest way to implement this is to replace that call to `traverse` with a call to a `mapM'` (maybe called `importsMapM`) that will reimplement traverse in all other cases except for `ImportAlt`.

Reply:
Oh, never mind.  Now I understand the problem you are describing.  The issue is an expression like this:

```haskell
App (ImportAlt a b) c
```

Once it pattern matches on `App` and doesn't find `ImportAlt` or `Embed` then it skips straight down to the `Embed` constructors at the leaves of the tree

I think the correct thing to do here is to just spell out the other cases instead of relying on `traverse` + `join`

Reply:
Speaking of which, I should probably simplify the AST a little bit to group operators and built-ins under one constructor each so that functions like these are easier to write

Reply:
Yes, +1 on simplifying the AST with groupings